### PR TITLE
Fixing DataGridView's scrollbars shouldn't include their control type in their accessible names

### LINKED ISSUE #2965: DataGridView's scrollbars shouldn't include their control type in their accessible names
**Steps to reproduce:**

1. Create a new WinForms .NET 4.8 app in VS.
2. Add a DataGridView which shows a horizontal scroll bar when run.
3. Run the app and have the Accessibility Insights for Windows tool examine the UI.

**Result**: The accessible name of the scrollbar is "Horizontal Scroll Bar", and so given that this includes the same (case insensitive) text as its localized control type of "scroll bar", Accessibility Insights reports this as an error.

**Expected**: A control shouldn't include its control type in its accessible name, and so the accessible name should be "Horizontal". (It's a similar case for the vertical scroll bar too.)

from @GuyBark: I spent a while trying to figure out how the accessible name is ending up as "Horizontal Scroll Bar". In https://referencesource.microsoft.com/#System.Windows.Forms/winforms/Managed/System/WinForms/DataGridView.cs, it sets this.horizScrollBar.AccessibleName to SR.GetString(SR.DataGridView_AccHorizontalScrollBarAccName), and in the code I found, that's "Horizontal" as it should be. But it seems that by the time a UIA client gets supplied the name, its "Horizontal Scroll Bar".

ported from AzDo 1071939

### PR DESCRIPTION
**Pull request by Ryuugamine**
Fixes #2965

## Proposed changes

- Added Accessible objects with a correct accessible name for ScrollBar classes
- Added unit tests

## Customer Impact

- With the fix available customer will be able to get correct information about the control using accessibility client apps.

## Regression? 

- No

## Risk

- Low

## Screenshots 
image

### Before

- "Vertical scroll bar" scroll bar

### After

- "Vertical" scroll bar

## Test methodology 

- Unit tests

## Accessibility testing  
- Inspect and Narrator

 

## Test environment(s) 

- DotNet version: 5.0.0-preview.4.20180.8

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/3038)

### CODE REVIEW

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.cs [hunk 1]
```diff
@@ -748,5 +748,79 @@ protected override void WndProc(ref Message m)
...
+            }
+
+            internal override void SetValue(double newValue)
+            {
+                throw new InvalidOperationException("Scroll Bar is read-only.");
+            }
+
+            internal override double LargeChange => double.NaN;
+
+            internal override double Maximum => OwningScrollBar?.Maximum ?? double.NaN;
```
Comment (hughbe) [inline src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.cs:815]: why would `OwningScrollBar` not be null?

File: src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/ScrollBarAccessibilitObjectTests.cs [hunk 2]
```diff
@@ -0,0 +1,28 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace System.Windows.Forms.Tests.AccessibleObjects
+{
+    public class ScrollBarAccessibilitObjectTests : IClassFixture<ThreadExceptionFixture>
```
Comment (hughbe) [inline src/System.Windows.Forms/tests/UnitTests/AccessibleObjects/ScrollBarAccessibilitObjectTests.cs:10]: nit: name of the file missing `y`

**Comment by RussKie** [comment]
Please rebase on top of the latest

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.AccessibleObject.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.AccessibleObject.cs:1]: ditto

File: src/System.Windows.Forms/src/System/Windows/Forms/VScrollBar.AccessibleObject.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/VScrollBar.AccessibleObject.cs:1]: ditto

File: src/System.Windows.Forms/tests/UnitTests/HScrollBarAccessibleObjectTests.cs [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace System.Windows.Forms.Tests
+{
+    public class HScrollBarAccessibleObjectTests : IClassFixture<ThreadExceptionFixture>
```
Comment (RussKie) [inline src/System.Windows.Forms/tests/UnitTests/HScrollBarAccessibleObjectTests.cs:14]: https://github.com/dotnet/winforms/blob/master/Documentation/testing.md#naming
```suggestion
    public class HScrollBar_HScrollBarAccessibleObjectTests : IClassFixture<ThreadExceptionFixture>
```

File: src/System.Windows.Forms/tests/UnitTests/ScrollBarAccessibilityObjectTests.cs [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Xunit;
+using static Interop.UiaCore;
+
+namespace System.Windows.Forms.Tests
+{
+    public class ScrollBarAccessibilityObjectTests : IClassFixture<ThreadExceptionFixture>
```
Comment (RussKie) [inline src/System.Windows.Forms/tests/UnitTests/ScrollBarAccessibilityObjectTests.cs:15]: ditto

File: src/System.Windows.Forms/tests/UnitTests/VScrollBarAccessibleObjectTests.cs [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace System.Windows.Forms.Tests
+{
+    public class VScrollBarAccessibleObjectTests : IClassFixture<ThreadExceptionFixture>
```
Comment (RussKie) [inline src/System.Windows.Forms/tests/UnitTests/VScrollBarAccessibleObjectTests.cs:14]: ditto

File: src/System.Windows.Forms/tests/UnitTests/HScrollBarAccessibleObjectTests.cs [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    {
+        [WinFormsFact]
```
Comment (RussKie) [inline src/System.Windows.Forms/tests/UnitTests/HScrollBarAccessibleObjectTests.cs:16]: Test for `HScrollBarAccessibleObject(null)`?

File: src/System.Windows.Forms/tests/UnitTests/ScrollBarAccessibilityObjectTests.cs [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    {
+        [WinFormsFact]
+        public void ScrollBarAccessibleObject_Ctor_Default()
```
Comment (RussKie) [inline src/System.Windows.Forms/tests/UnitTests/ScrollBarAccessibilityObjectTests.cs:18]: ditto

File: src/System.Windows.Forms/tests/UnitTests/VScrollBarAccessibleObjectTests.cs [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+    {
+        [WinFormsFact]
+        public void VScrollBarAccessibleObject_Ctor_Default()
```
Comment (RussKie) [inline src/System.Windows.Forms/tests/UnitTests/VScrollBarAccessibleObjectTests.cs:17]: ditto

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.AccessibleObject.cs [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+            private ScrollBar _owningScrollBar;
+
+            internal ScrollBarAccessibleObject(ScrollBar owner) : base(owner)
+            {
+                _owningScrollBar = owner;
+            }
+
+            internal override bool IsIAccessibleExSupported() => true;
+
+            internal override bool IsPatternSupported(UiaCore.UIA patternId)
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.AccessibleObject.cs:27]: Tests?

File: src/System.Windows.Forms/tests/UnitTests/ScrollBarAccessibilityObjectTests.cs [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+            Assert.False(scrollBar.IsHandleCreated);
+            var scrollBarAccessibleObject = new ScrollBar.ScrollBarAccessibleObject(scrollBar);
+            object value = scrollBarAccessibleObject.GetPropertyValue((UIA)propertyID);
+
+            Assert.Equal(expected, value);
+            // TODO: ControlAccessibleObject shouldn't force handle creation, tracked in https://github.com/dotnet/winforms/issues/3062
+            Assert.True(scrollBar.IsHandleCreated);
+        }
+
+        private class SubScrollBar : ScrollBar
```
Comment (RussKie) [inline src/System.Windows.Forms/tests/UnitTests/ScrollBarAccessibilityObjectTests.cs:52]: Why is this needed?

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.AccessibleObject.cs [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+            internal override object? GetPropertyValue(UiaCore.UIA propertyID)
+                => propertyID switch
+                {
+                    UiaCore.UIA.NamePropertyId => Name,
+                    UiaCore.UIA.AutomationIdPropertyId => _owningScrollBar.Name,
+                    UiaCore.UIA.ControlTypePropertyId => UiaCore.UIA.ScrollBarControlTypeId,
+                    UiaCore.UIA.IsKeyboardFocusablePropertyId => true,
+                    UiaCore.UIA.IsValuePatternAvailablePropertyId => IsPatternSupported(UiaCore.UIA.ValuePatternId),
+                    _ => base.GetPropertyValue(propertyID)
+                };
```
Comment (vladimir-krestov) [inline src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.AccessibleObject.cs:46]: Please check if we need this for fixing the issue.
(I'm about the whole implementation and `_owningScrollBar` too)
If it is not necessary, please remove this to avoid adding regression issues.

File: src/System.Windows.Forms/tests/UnitTests/ScrollBarAccessibilityObjectTests.cs (hunk 12 above)
Comment (Ryuugamine) [reply src/System.Windows.Forms/tests/UnitTests/ScrollBarAccessibilityObjectTests.cs:52]: ScrollBar is an abstract class and we can't test it without extending

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.AccessibleObject.cs (hunk 13 above)
Comment (Ryuugamine) [reply src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.AccessibleObject.cs:46]: Current issue was fixed at master branch but we still need this UIA implementation in regard of Caps Navigation issue #2164

**Comment by Ryuugamine** [comment]
Test team approved this changes
